### PR TITLE
Send master connection string, never dismiss the connection.

### PR DESCRIPTION
**Pull request by ktsakalozos**
Not very happy with the fact that I do not dismiss the relation when all units have departed...

### CODE REVIEW

File: requires.py [hunk 1]
```diff
@@ -32,16 +32,19 @@ def changed(self):
         conv.toggle_state('{relation_name}.ready',
                           active=self.is_spark_started())
 
-    @hook('{requires:spark}-relation-departed')
-    def departed(self):
-        conv = self.conversation()
-        conv.remove_state('{relation_name}.joined')
-        conv.remove_state('{relation_name}.ready')
```
Comment (johnsca) [inline requires.py:39]: Leaving the states indefinitely will cause other issues.  The best interim work-around I can think of is to test against the remaining units, i.e.:

``` python
    @hook('{requires:spark}-relation-departed')
    def departed(self):
        conv = self.conversation()
        if len(conv.units) == 1: # last remaining unit departing
            conv.remove_state('{relation_name}.joined')
            conv.remove_state('{relation_name}.ready')
```

It's not ideal to have to do this in every interface layer, but I don't see a way around it right now.  Probably the framework needs to handle the `-broken` hook and automatically clean up _all_ reactive relation states, but there was an issue with implementing that which I hadn't figured out how to deal with yet.

Comment (ktsakalozos) [inline requires.py:39]: Ok, fixed and tested it works as expected.

File: requires.py [hunk 2]
```diff
@@ -32,11 +32,11 @@ def changed(self):
         conv.toggle_state('{relation_name}.ready',
                           active=self.is_spark_started())
 
-    @hook('{requires:spark}-relation-departed')
+    @hook('{provides:spark}-relation-departed')
```
Comment (kwmonroe) [inline requires.py]: i think you mean `requires` instead of `provides` here, since we're in requires.py.

Comment (ktsakalozos) [inline requires.py]: Thank you for spotting that. Should be fixed in the next commit.

File: requires.py [hunk 3]
```diff
@@ -45,3 +46,12 @@ def get_private_ip(self):
...
         return conv.get_remote('rest_port')
+
+    def get_master_info(self):
+        conv = self.conversation()
+        data={
+            'connection_string': conv.get_remote('connection_string'),
+            'master': conv.get_remote('master'),
+        }
+        return data
+
```
Comment (pengale) [inline requires.py]: This blank line at the end makes the linter fail, when run via "make lint" on the built charm. At least it's an easy fix :-)

**Comment by pengale** [comment]
Code looks good. Not sure how to test it, though (see my comments on the related PRs).

**Comment by kwmonroe** [comment]
LGTM

**Pull request merged**